### DISCUSSION TITLE
How to modify the code to adapt to the latest version of Zenoh?

### DISCUSSION THREAD

**Discussion by stevebobaaa** [eclipse-zenoh/roadmap#102] [Q&A]
The following code is from the old version of Zenoh, and some of the packages used in it are no longer usable. How can I modify it?

Thank you so much！

```
conf = zenoh.Config()
basekey = "/sdn/zone1"
zenoh.init_logger()
session = zenoh.open(conf)
zones = ["zone2","zone3"]

sub1 = session.subscribe("/sdn/host-pkt/**",topology_update,reliability=Reliability.Reliable, mode=SubMode.Push)

store = {}
sub = session.subscribe(basekey + "/hosts/**", listener, reliability=Reliability.Reliable, mode=SubMode.Push)
queryable = session.queryable(basekey + "/hosts/**",STORAGE, query_handler)
store_bs = {}
sub_bs = session.subscribe(basekey + "/BS/**", listener_bs, reliability=Reliability.Reliable, mode=SubMode.Push)
queryable_bs = session.queryable(basekey + "/BS/**",  STORAGE,query_handler_bs)
boosted = []
paths_cache = {}
mac_to_port = {}
modes=["stdnetwork","prioritynetwork"]
mode = modes[0]
net=nx.DiGraph()
fast_net = None
slow_net = None
nodes = {}
links = {}
switches = {}
no_of_nodes = 0
no_of_links = 0
i=0
zone = 1
#fast_links = []
fast_links = [(to_dpid(2),to_dpid(4)),(to_dpid(4),to_dpid(2)),(to_dpid(4),to_dpid(1)),(to_dpid(1),to_dpid(4)),("00:00:00:00:00:01",to_dpid(1)),(to_dpid(1),"00:00:00:00:00:01"),("00:00:00:00:00:02",to_dpid(2)),(to_dpid(2),"00:00:00:00:00:01")]
known_hosts = {"00:00:00:00:00:01":{"zone":1, "ip":"10.0.0.1"},"00:00:00:00:00:02":{"zone":1, "ip":"10.0.0.2"}} #mac:zone
border_switch=[3,4]
session.put(basekey + "/BS/2",json.dumps([{"via":"3","from":"1"},{"via":"4","from":"1"}])) #to zone via dpid from zone s
border_gw = {2:'gateway2'} #to zone %d use gw %d
flows = {}
flag = 0
flag_flows = 0

target = Target.BestMatching()

br_delay = 0
host_delay = 0
comm_type = 0
```

**Comment by gabrik** [comment]
Hi @stevebobaaa,

Which version of Zenoh are you using?

In Zenoh website you can find a section with migration guides from older version of the APIs to new ones: https://zenoh.io/docs/migration

**Reply by stevebobaaa** [reply]
The Zenoh used in the code is probably a certain version in 2022, it was a nightly release. 
The link you gave me seems to have expired. After opening it, it was 404 not found. May I ask if it's a problem with my network?

**Reply by stevebobaaa** [reply]
The code has imported SubMode, Target, and so on, but it seems that the current version has become invalid.

**Reply by gabrik** [reply]
Ops sorry: this should be the valid link: https://zenoh.io/docs/migration/migrationguide-python-v0.5.x-v0.6.x/ (API should be stable from 0.6)

Can you do a `pip3 list | grep zenoh` just to be sure which version you are using currently

**Reply by stevebobaaa** [reply]
eclipse-zenoh          0.10.0rc0
This is the currently used version of Zenoh.

**Reply by stevebobaaa** [reply]
https://github.com/Jiarre/SDN-distributed-controllers/blob/main/hybrid/controllerz1.py
This is the link to the complete code. Do you have time to take a look? Running under the current Zenoh version will result in error messages.

**Reply by gabrik** [reply]
The easiest fix is already to remove the leading `/` from the KeyExpressions.

Then I suggest having a look at the examples: https://github.com/eclipse-zenoh/zenoh-python/tree/master/examples
and the docs: https://zenoh-python.readthedocs.io/en/0.10.0-rc/

It should not be a lot of changes

**Reply by stevebobaaa** [reply]
https://zenoh-python.readthedocs.io/en/0.5.0-b8/zenoh-net-api.html
I have found some interfaces used in the code in this link, such as SubMode. How can I convert them into ones that can be used in the current version?

**Reply by stevebobaaa** [reply]
Or can I use a Zenoh of 0.5.0-b8? How can I install it?

**Reply by gabrik** [reply]
We really discouraged using such an old version of Zenoh, I think porting the code is the better solution.

**Reply by stevebobaaa** [reply]
OK, of course, porting the code is the best way, but the https://zenoh.io/docs/migration/migrationguide-python-v0.5.x-v0.6.x/ link says that Zenoh net has been removed. Which APIs can be replaced in the new version of Zenoh? And the Discord in the link is no longer available. What should I do?
Thank you very much!

**Reply by gabrik** [reply]
Zenoh-Net has been integrated in the Zenoh API.
Thank you for poiting out that the discord link does not work, here the correct one: https://discord.gg/Z9f8fKydKK 
We are going to update the one in the webpage
